### PR TITLE
[WIP][DTensor] Let DTensor apply static padding on local tensor if needed

### PR DESCRIPTION
**Pull request by wz337**
Fixes #ISSUE_NUMBER

In this prototype, we demonstrate how to derive the current shard size given a global shape, device_mesh, and placements. 
With this, we calculate a `unpadded_local_shard_size`:
1. `unpadded_local_shard_size` = None if no padding is required for current shard 
2. `unpadded_local_shard_size` matches the shape of local tensor without padding, which is < dtensor._local_tensor.shape.

The benefit of having this property is that we don't need to do `torch.narrow(input, dim, start, length)` on each tensor dim, if more than one dim needs padding. 

We will eventually break up the utility function and make `unpadded_local_shard_size` a cached property. This PR is just to demonstrate the idea and show how `unpadded_local_shard_size` is derived. 

As expected, lots of test cases would break.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/_tensor/api.py [hunk 1]
```diff
@@ -70,7 +70,16 @@ def forward(  # type: ignore[override]
     ):
         ctx.dtensor_spec = input._spec
         ctx.grad_placements = grad_placements
-        local_tensor = input._local_tensor
+
+        cur_unpadded_shard_size = ctx.dtensor_spec.unpadded_local_shard_size
```
Comment (wz337) [inline torch/distributed/_tensor/api.py:74]: the MYPY attribute error is expected, as we didn't add this properly yet.

File: torch/distributed/_tensor/api.py [hunk 2]
```diff
@@ -252,6 +261,91 @@ def __new__(
...
+                # so we don't need to re-calculate this during backward of _ToTorchTensor.
+                spec.unpadded_local_shard_size = None
+                return local_tensor, spec
+
+            # 5. create a padded local tensor and copy local tensor into padded local tensor
+            spec.unpadded_local_shard_size = (
+                cur_unpadded_shard_size  # make this a cached property in DTensorSpec
+            )
+            padded_local_tensor = local_tensor.new_zeros(full_shard_size)
+            slices = [
```
Comment (wanchaol) [inline torch/distributed/_tensor/api.py:338]: why we still need to calculate the slice here? IMO cur_unpadded_shard_size should just be directly used as slice? if not we should just return the `slices` directly?

File: torch/distributed/_tensor/api.py [hunk 3]
```diff
@@ -252,6 +261,91 @@ def __new__(
...
+                slice(0, unpadded_dim_size)
+                for unpadded_dim_size in cur_unpadded_shard_size
+            ]
+            padded_local_tensor[slices].copy_(local_tensor)
```
Comment (wanchaol) [inline torch/distributed/_tensor/api.py:342]: IMO this kind of "allocating new stoarage + copy the local tensor content" logic should only live as a util and only be used in `from_local` logic. 

And actually I don't really know what's the advantage of making a new_zeros and then copying over, instead of calling the `pad_tensor` directly, which is more neat? Could you elaborate why this approach?

**Review by wanchaol — COMMENTED** [review]
Please see inlined comments

File: torch/distributed/_tensor/api.py (hunk 3 above)
Comment (wz337) [reply torch/distributed/_tensor/api.py:342]: From what I read, `torch.nn.functional.pad` returns a new tensor. So from the padding perspective, both `torch.nn.functional.pad` and making a new_zeros and copying over both return a new tensor. The difference is during unpadding (unpad each tensor dimension vs unpad is done all at once with the slices generated from cur_unpadded_shard_size). 

However, since we already have the slices from cur_unpadded_shard_size, I just thought we can utilize it for padding as well. In terms of readability, making a new_zeros and copying over is more readable IMO.

File: torch/distributed/_tensor/api.py [hunk 4]
```diff
@@ -252,6 +261,91 @@ def __new__(
             requires_grad=requires_grad,
         )
 
+        # The utility function needs to be broken up and should live somewhere else instead of being in the __new__,
+        # but this logic needs to happen here at runtime before we assign the local tensor to the wrapper tensor.
```
[the remark this answers is not in the document]
Comment (wz337) [reply torch/distributed/_tensor/api.py:265]: Ye. Thanks for the correction. This is specifically for the FSDP2 usage currently. FSDP either uses `DTensor` or `DTensor.from_local()` right now. If we are using `distribute_tensor` API, `__new__` should already receive a padded tensor.

File: torch/distributed/_tensor/api.py (hunk 2 above)
Comment (wz337) [reply torch/distributed/_tensor/api.py:338]: This is just to generate the slice index for us. 
If we know the tensor dim, we can simply do something like, tensor[:2, :3, :4], but here we are dynamically generating the slices based on cur_unpadded_shard_size. 

If we need to add a property to DTensorSpec, cur_unpadded_shard_size would be more readable then this slice object.

File: torch/distributed/_tensor/api.py [hunk 5]
```diff
@@ -252,6 +261,91 @@ def __new__(
...
+                    tensor_dim_size / num_chunks_by_dim[tensor_dim]
+                )
+                num_full_chunks = math.floor(tensor_dim_size / full_chunk_size)
+                tail_chunk_size = tensor_dim_size % full_chunk_size
+                full_shard_size.append(full_chunk_size)
+
+                # Currently we assume the shard id is to equal to the global rank.
+                # Note that this might not be true for 2D (strided sharding + global_rank)
+                # and 3D (strided sharding + 2D local rank).
+                cur_chunk = mesh.get_rank()
```
[the remark this answers is not in the document]
Comment (wz337) [reply torch/distributed/_tensor/api.py:297]: Ah. Thanks for the correction.
